### PR TITLE
AP_Frsky_Telem: use ringbuffer for Frsky message queueing

### PR DESCRIPTION
**Pull request by floaledm**
As suggested by @tridge to use a ringbuffer for Frsky message queueing

### CODE REVIEW

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 1]
```diff
@@ -435,16 +437,16 @@ uint32_t AP_Frsky_Telem::get_next_msg_chunk(void)
 {
     if (_msg_chunk.repeats == 0) {
         _msg_chunk.chunk = 0;
-        uint8_t character = _msg.data[_msg.sent_idx].text[_msg_chunk.char_index++];
+        uint8_t character = _statustext_queue[0]->text[_msg_chunk.char_index++];
```
Comment (magicrub) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Does there need to be an empty check here so we don't use [0] which returns null on an empty buffer?

**Comment by dgrat** [comment]
What would be the advantage in comparison to a LIFO?

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 1 above)
Comment (floaledm) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Good point Tom. The empty check is performed right before get_next_msg_chunk() is called...

**Comment by floaledm** [comment]
The ringbuffer is also used for MAVLink. With a LIFO, there's a significant chance some of the messages would be transmitted out of order.

**Comment by magicrub** [comment]
For status messages, the order is important

**Comment by magicrub** [comment]
@floaledm PROTIP: branch names other than "dev#" are useful when others review branches :)

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 1 above)
Comment (magicrub) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: I see that, but what happens when this function gets reused? The check should be in the function, it's not wise to have other functions sanity check contents of other functions. That's a future bug in the making..

**Comment by magicrub** [comment]
I think it needs a little more index bound checking (within the function doing the indexing) to be more robust but it's better than what's currently in so I merged it.

**Comment by floaledm** [comment]
Thanks Tom. I could add another _statustext_queue.empty() check at the beginning of get_next_msg_chunk() just to be on the safe side. What do you think?

**Comment by magicrub** [comment]
I was going to rewrite it to do that in the other function but the calling function behaves differently if it's full/empty so I didn't want to change anything. Adding a redundant check is currently redundant, is there a way to change the logic in the previous function to not need the check? If it needs it, then I guess a double check is better than what we have now.

**Pull request closed without merging**